### PR TITLE
Read KVP values in binary mode

### PR DESCRIPTION
**Pull request by longlimsft**
The KVP data is stored in binary format. The agent should read them in binary
mode. This fixes an issue with python3 incorrectly reading values from KVP.

## Description

Issue # 

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made and Travis.CI is passing.

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/common/rdma.py [hunk 1]
```diff
@@ -94,13 +94,17 @@ def get_rdma_version(self):  # pylint: disable=R1710
...
-        f = open(driver_info_source)  # pylint: disable=C0103
+        f = open(driver_info_source, "rb")  # pylint: disable=C0103
         while True:
             key = f.read(kvp_key_size)
             value = f.read(kvp_value_size)
             if key and value:
-                key_0 = key.split("\x00")[0]
-                value_0 = value.split("\x00")[0]
+                key_0 = key.partition(b"\x00")[0]
+                if key_0 :
```
Comment (pgombar) [inline azurelinuxagent/common/rdma.py:103]: nit: please remove extra whitespace

File: azurelinuxagent/common/rdma.py [hunk 2]
```diff
@@ -94,13 +94,17 @@ def get_rdma_version(self):  # pylint: disable=R1710
...
+                    key_0 = key_0.decode()
+                value_0 = value.partition(b"\x00")[0]
+                if value_0 :
```
Comment (pgombar) [inline azurelinuxagent/common/rdma.py:106]: nit: please remove extra whitespace

**Review by pgombar — COMMENTED** [review]
Thank you for the fix. Could you please add a unit test for the changes? We need to support both Python2 and Python3 at the same time and make sure this is not a breaking change.

**Comment by longlimsft** [comment]
@pgombar, this feature can only be tested in a HPC VM that reports ND version or SRIOV IB information over KVP. We have tested on both Python2 and Python3.

When a VM is deployed to a non-HPC size, this code path is not exercised.

**Comment by pgombar** [comment]
> @pgombar, this feature can only be tested in a HPC VM that reports ND version or SRIOV IB information over KVP. We have tested on both Python2 and Python3.
> 
> When a VM is deployed to a non-HPC size, this code path is not exercised.

Thanks for confirming.

**Comment by pgombar** [comment]
Running end-to-end automation: https://tuxgold.corp.microsoft.com/job/AzLinux/job/DungeonCrawler/view/WALinuxAgent/job/DungeonCrawler.pagombar2/15/

**Review by pgombar — DISMISSED** [review]
End-to-end automation ok (no new issues).

**Review by kevinclark19a — DISMISSED** [review]

File: azurelinuxagent/common/rdma.py [hunk 3]
```diff
@@ -94,13 +94,17 @@ def get_rdma_version(self):  # pylint: disable=R1710
             logger.error(error_msg % driver_info_source)
             return
 
-        f = open(driver_info_source)  # pylint: disable=C0103
+        f = open(driver_info_source, "rb")  # pylint: disable=C0103
```
Comment (larohra) [inline azurelinuxagent/common/rdma.py:97]: NIT: Please fix the pylint warning 
`# pylint: disable=C0103`

**Review by narrieta — DISMISSED** [review]

File: azurelinuxagent/common/rdma.py [hunk 4]
```diff
@@ -94,13 +94,17 @@ def get_rdma_version(self):  # pylint: disable=R1710
...
-                key_0 = key.split("\x00")[0]
-                value_0 = value.split("\x00")[0]
+                key_0 = key.partition(b"\x00")[0]
+                if key_0:
+                    key_0 = key_0.decode()
+                value_0 = value.partition(b"\x00")[0]
+                if value_0:
+                    value_0 = value_0.decode()
                 if key_0 == "NdDriverVersion":
                     f.close()
```
Comment (larohra) [inline azurelinuxagent/common/rdma.py:109]: This comment does not directly related to your changes, but could you please move the file closing to either a `finally` clause or move the file operations to under a `with` clause. This is to ensure that the file always gets closed even if there is some exception thrown while processing the data. 

Sample usage: https://cmdlinetips.com/2016/01/opening-a-file-in-python-using-with-statement/

Thanks!

**Review by larohra — CHANGES_REQUESTED** [review]

**Review by larohra — APPROVED** [review]

**Review by narrieta — APPROVED** [review]

**Comment by pgombar** [comment]
Running automation again: https://tuxgold.corp.microsoft.com/job/AzLinux/job/DungeonCrawler/view/WALinuxAgent/job/DungeonCrawler.pagombar2/16/

**Pull request merged**